### PR TITLE
Implement copy(), copy/deepcopy and __eq__ for jacobians, observations, gmix, as well as get_data for jacobians

### PR DESCRIPTION
**Pull request by esheldon**

### CODE REVIEW

**Comment by esheldon** [comment]
while I'm at it I'm going to add some more of these for other ngmix data types

**Review by beckermr — CHANGES_REQUESTED** [review]

**Comment by esheldon** [comment]
I'm stumped currently on this infinite recursion

**Comment by beckermr** [comment]
Hmmmm.

The docs say:

```
In order for a class to define its own copy implementation, it can define special methods __copy__() and __deepcopy__(). The former is called to implement the shallow copy operation; no additional arguments are passed. The latter is called to implement the deep copy operation; it is passed one argument, the memo dictionary. If the __deepcopy__() implementation needs to make a deep copy of a component, it should call the deepcopy() function with the component as first argument and the memo dictionary as second argument. The memo dictionary should be treated as an opaque object.
```

Maybe we need to the memo dict for our object before calling copy and just return any copy we have already made?

We also might want to use the memo dict properly per the docs.

**Comment by esheldon** [comment]
the id won't have changed, so we need to replace it in memo

**Comment by esheldon** [comment]
I mean if, say the arrays change

File: ngmix/observation.py [hunk 1]
```diff
@@ -979,6 +1054,34 @@ def get_s2n_sums(self):
...
+        for obslist in self:
+            new_mbobs.append(obslist.copy(memo=memo))
+
+        return new_mbobs
+
+    def __copy__(self):
+        return self.copy()
+
+    def __deepcopy__(self, memo):
+        result = self.copy()
```
Comment (beckermr) [inline ngmix/observation.py:1072]:
```suggestion
        result = self.copy(memo=memo)
```

**Comment by beckermr** [comment]
Huzzah!

**Review by beckermr — APPROVED** [review]

**Comment by beckermr** [comment]
nice work!

**Comment by esheldon** [comment]
thanks for the excellent review

**Pull request merged**